crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,261 @@
+/// differently, as those assets are extracted to render world resources and not entities.
+///
+/// [`PipelinedRenderingPlugin`]: crate::pipelined_rendering::PipelinedRenderingPlugin
+#[derive(Default)]
+pub struct WorldSyncPlugin;
+
+impl Plugin for WorldSyncPlugin {
+    fn build(&self, app: &mut bevy_app::App) {
+        app.init_resource::<PendingSyncEntity>();
+        app.observe(
+            |trigger: Trigger<OnAdd, SyncRenderWorld>, mut pending: ResMut<PendingSyncEntity>| {
+                pending.push(EntityRecord::Added(trigger.entity()));
+            },
+        );
+        app.observe(
+            |trigger: Trigger<OnRemove, SyncRenderWorld>,
+             mut pending: ResMut<PendingSyncEntity>,
+             query: Query<&RenderEntity>| {
+                if let Ok(e) = query.get(trigger.entity()) {
+                    pending.push(EntityRecord::Removed(e.id()));
+                };
+            },
+        );
+    }
+}
+/// Marker component that indicates that its entity needs to be Synchronized to the render world
+///
+/// NOTE: This component should persist throughout the entity's entire lifecycle.
+/// If this component is removed from its entity, the entity will be despawned.
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[reflect[Component]]
+#[component(storage = "SparseSet")]
+pub struct SyncRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug, Copy)]
+/// Marker component added on the main world entities that are synced to the Render World in order to keep track of the corresponding render world entity
+pub struct RenderEntity(Entity);
+impl RenderEntity {
+    #[inline]
+    pub fn id(&self) -> Entity {
+        self.0
+    }
+}
+
+#[derive(Component, Deref, Clone, Debug)]
+/// Marker component added on the render world entities to keep track of the corresponding main world entity
+pub struct MainEntity(Entity);
+impl MainEntity {
+    #[inline]
+    pub fn id(&self) -> Entity {
+        self.0
+    }
+}
+
+// marker component that indicates that its entity needs to be despawned at the end of every frame.
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[component(storage = "SparseSet")]
+pub struct TemporaryRenderEntity;
+
+pub(crate) enum EntityRecord {
+    // When an entity is spawned on the main world, notify the render world so that it can spawn a corresponding entity. This contains the main world entity
+    Added(Entity),
+    // When an entity is despawned on the main world, notify the render world so that the corresponding entity can be despawned. This contains the render world entity.
+    Removed(Entity),
+}
+
+// Entity Record in MainWorld pending to Sync
+#[derive(Resource, Default, Deref, DerefMut)]
+pub(crate) struct PendingSyncEntity {
+    records: Vec<EntityRecord>,
+}
+
+pub(crate) fn entity_sync_system(main_world: &mut World, render_world: &mut World) {
+    main_world.resource_scope(|world, mut pending: Mut<PendingSyncEntity>| {
```

Comment:
This function could use some splitting up into closures or smaller functions to reduce the nesting 😬